### PR TITLE
Make IP_PKTINFO optional.

### PR DESCRIPTION
**Pull request by ahedberg**
For #5027

### CODE REVIEW

File: src/core/iomgr/socket_utils_common_posix.c [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /*
  *
- * Copyright 2015, Google Inc.
+ * Copyright 2016, Google Inc.
```
Comment (nicolasnoble) [inline src/core/iomgr/socket_utils_common_posix.c]: The file existed in 2015, we need to stick with the (C) from there. So 2015-2016.

Comment (ahedberg) [inline src/core/iomgr/socket_utils_common_posix.c]: Done.

File: src/core/iomgr/socket_utils_common_posix.c [hunk 2]
```diff
@@ -89,6 +89,26 @@ int grpc_set_socket_no_sigpipe_if_possible(int fd) {
 #endif
 }
 
+int grpc_set_socket_ip_pktinfo_if_possible(int fd) {
+#ifdef GPR_HAVE_IP_PKTINFO
+  int get_local_ip = 1;
+  return 0 == setsockopt(fd, IPPROTO_IP, IP_PKTINFO, &get_local_ip,
+                         sizeof(get_local_ip));
+#else
+  return 1;
```
Comment (nicolasnoble) [inline src/core/iomgr/socket_utils_common_posix.c]: Please add `(void)fd` to tell some older compilers that the fact the `fd` variable isn't being referenced from that codepath isn't a problem.

File: src/core/iomgr/socket_utils_common_posix.c [hunk 3]
```diff
@@ -89,6 +89,26 @@ int grpc_set_socket_no_sigpipe_if_possible(int fd) {
...
+#endif
+}
+
+int grpc_set_socket_ipv6_recvpktinfo_if_possible(int fd) {
+#ifdef GPR_HAVE_IPV6_RECVPKTINFO
+  int get_local_ip = 1;
+  return 0 == setsockopt(fd, IPPROTO_IPV6, IPV6_RECVPKTINFO, &get_local_ip,
+                         sizeof(get_local_ip));
+#else
+  return 1;
```
Comment (nicolasnoble) [inline src/core/iomgr/socket_utils_common_posix.c]: Ditto.

**Comment by nicolasnoble** [comment]
You'll want to do a `git merge master` from upstream in order to take #5964 into account.

File: src/core/iomgr/socket_utils_common_posix.c (hunk 2 above)
Comment (ahedberg) [inline src/core/iomgr/socket_utils_common_posix.c]: Done

File: src/core/iomgr/socket_utils_common_posix.c (hunk 3 above)
Comment (ahedberg) [inline src/core/iomgr/socket_utils_common_posix.c]: Done

**Comment by ahedberg** [comment]
Merged.

**Comment by nicolasnoble** [comment]
Alright, everything looks good from my side. Since you've touched the iOS side a tiny bit, I'd like to see the result of the travis tests to merge it, but otherwise, LGTM.

**Pull request merged**